crates/rpc/src/server/api.rs

```diff
@@ -88,6 +91,54 @@ impl RpcService {
 
         Self { store, block_producer }
     }
+
+    /// Fetches the genesis block header from the store.
+    ///
+    /// Automatically retries until the store connection becomes available.
+    pub async fn get_genesis_header_with_retry(&self) -> anyhow::Result<BlockHeader> {
+        let mut retry_counter = 0;
+        loop {
+            let result = self
+                .get_block_header_by_number(
+                    GetBlockHeaderByNumberRequest {
+                        block_num: Some(BlockNumber::GENESIS.as_u32()),
+                        include_mmr_proof: None,
+                    }
+                    .into_request(),
+                )
+                .await;
+
+            match result {
+                Ok(header) => {
+                    let header = header
+                        .into_inner()
+                        .block_header
+                        .context("response is missing the header")?;
+                    let header =
+                        BlockHeader::try_from(header).context("failed to parse response")?;
+
+                    return Ok(header);
+                },
+                Err(err) if err.code() == tonic::Code::Unavailable => {
+                    // exponential backoff with base 500ms and max 30s
+                    let backoff = Duration::from_millis(500)
+                        .saturating_mul(1 << retry_counter)
+                        .min(Duration::from_secs(30));
+
+                    tracing::warn!(
+                        ?backoff,
+                        %retry_counter,
+                        %err,
+                        "connection failed while subscribing to the mempool, retrying"
+                    );
+
+                    retry_counter += 1;
```

Comment:
Does this loop forever until it succeeds? Should we add a limit and error out instead? This would be something handled by the utility fn I mentioned  [here](https://github.com/0xMiden/miden-node/pull/1084/files#r2205996865).

Reply:
One more thought on this - with the loop + WARN approach we would need individual alerts for each such situation. Whereas crash loops would be alerted on by my general alerts that are not so implementation-aware.